README.md
# Serverless Alexa Events

[![Travis build status](http://img.shields.io/travis/rajington/serverless-alexa-plugin.svg?style=flat)](https://travis-ci.org/rajington/serverless-alexa-plugin)
[![Code Climate](https://codeclimate.com/github/rajington/serverless-alexa-plugin/badges/gpa.svg)](https://codeclimate.com/github/rajington/serverless-alexa-plugin)
[![Test Coverage](https://codeclimate.com/github/rajington/serverless-alexa-plugin/badges/coverage.svg)](https://codeclimate.com/github/rajington/serverless-alexa-plugin)
[![Dependency Status](https://david-dm.org/rajington/serverless-alexa-plugin.svg)](https://david-dm.org/rajington/serverless-alexa-plugin)
[![devDependency Status](https://david-dm.org/rajington/serverless-alexa-plugin/dev-status.svg)](https://david-dm.org/rajington/serverless-alexa-plugin#info=devDependencies)

This plugins compiles the Alexa Skills Kit events to a CloudFormation resource.

It requires Serverless 1.0 or later.

## How it works

`Compile Alexa Skills Kit Events` hooks into the [`deploy:compileEvents`](/lib/plugins/deploy) lifecycle.

It loops over all functions which are defined in `serverless.yaml`. For each function that have Alexa
events defined, a lambda permission for the current function is created which makes is possible to invoke the
function when the skill is spoken.

Take a look at the [Event syntax examples](#event-syntax-examples) below to see how you can setup an Alexa Skills Kit event.

The resource is then merged into the `serverless.service.resources.Resources` section.

## Event syntax examples

Enable Alexa Skills Kit event:

```yaml
# serverless.yaml
functions:
    greet:
        handler: handler.hello
        events:
            - alexaSkillsKit
```

## TODO:

I do not believe Alexa Smart Home events are possible via CloudFormation, although they are supported via [command-line](http://docs.aws.amazon.com/cli/latest/reference/lambda/add-permission.html?highlight=event-source-token) and [SDK](https://github.com/aws/aws-sdk-js/blob/master/apis/lambda-2015-03-31.normal.json#L474).

It may require creating a separate policy to be linked, will have to investigate. The plugin accepts the value and supplies it to CloudFormation, but it currently fails.

Configure Alexa Smart Home event:

```yaml
# serverless.yaml
functions:
    greet:
        handler: handler.hello
        events:
            - alexaSmartHome: amzn1.ask.skill.12345678-1234-4234-8234-9234567890AB
```


CHANGELOG.md
### [0.0.1](https://github.com/rajington/serverless-alexa-plugin/releases/tag/v0.0.1)

- The first release


test/setup/setup.js
/* eslint-disable */

module.exports = function(root) {
  root = root ? root : global;
  root.expect = root.chai.expect;

  beforeEach(function() {
    // Using these globally-available Sinon features is preferrable, as they're
    // automatically restored for you in the subsequent `afterEach`
    root.sandbox = root.sinon.sandbox.create();
    root.stub = root.sandbox.stub.bind(root.sandbox);
    root.spy = root.sandbox.spy.bind(root.sandbox);
    root.mock = root.sandbox.mock.bind(root.sandbox);
    root.useFakeTimers = root.sandbox.useFakeTimers.bind(root.sandbox);
    root.useFakeXMLHttpRequest = root.sandbox.useFakeXMLHttpRequest.bind(root.sandbox);
    root.useFakeServer = root.sandbox.useFakeServer.bind(root.sandbox);
  });

  afterEach(function() {
    delete root.stub;
    delete root.spy;
    root.sandbox.restore();
  });
};


test/setup/browser.js
/* eslint-disable */

var mochaGlobals = require('./.globals.json').globals;

window.mocha.setup('bdd');
window.onload = function() {
  window.mocha.checkLeaks();
  window.mocha.globals(Object.keys(mochaGlobals));
  window.mocha.run();
  require('./setup')(window);
};


test/unit/serverless-alexa-plugin.js
import Serverless from 'serverless';
import ServerlessAlexaPlugin from '../../src/serverless-alexa-plugin';

describe('ServerlessAlexaPlugin', () => {
  let serverless;
  let alexaPlugin;

  beforeEach(() => {
    serverless = new Serverless();
    serverless.service.resources = { Resources: {} };
    alexaPlugin = new ServerlessAlexaPlugin(serverless);
    alexaPlugin.serverless.service.service = 'new-service';
  });

  describe('#constructor()', () => {
    it('should set the provider variable to "aws"', () =>
      expect(
        alexaPlugin.provider
      ).to.equal('aws'));
  });

  describe('#compileAlexaEvents()', () => {
    it('should throw an error if the resource section is not available', () => {
      alexaPlugin.serverless.service.resources.Resources = false;
      expect(() => alexaPlugin.compileAlexaEvents()).to.throw(Error);
    });

    it('should create corresponding resources when alexaSkillsKit events are given', () => {
      alexaPlugin.serverless.service.functions = {
        first: {
          events: [
            'alexaSkillsKit',
          ],
        },
      };

      alexaPlugin.compileAlexaEvents();

      expect(alexaPlugin.serverless.service
        .resources.Resources.firstAlexaEventPermission0.Type
      ).to.equal('AWS::Lambda::Permission');
    });

    it('should create corresponding resources when alexaSmartHome events are given', () => {
      const alexaAppId = 'amzn1.ask.skill.12345678-1234-4234-8234-9234567890AB';
      alexaPlugin.serverless.service.functions = {
        first: {
          events: [
            {
              alexaSmartHome: alexaAppId,
            },
          ],
        },
      };

      alexaPlugin.compileAlexaEvents();

      expect(alexaPlugin.serverless.service
        .resources.Resources.firstAlexaEventPermission0
        .Properties.EventSourceToken
      ).to.equal(alexaAppId);
    });

    it('should not create corresponding resources when ask events are not given', () => {
      alexaPlugin.serverless.service.functions = {
        first: {
          events: [],
        },
      };

      alexaPlugin.compileAlexaEvents();

      expect(
        alexaPlugin.serverless.service.resources.Resources
      ).to.deep.equal({});
    });
  });
});


gulpfile.js
/* eslint-disable */

const gulp = require('gulp');
const loadPlugins = require('gulp-load-plugins');
const del = require('del');
const glob = require('glob');
const path = require('path');
const isparta = require('isparta');
const webpack = require('webpack');
const webpackStream = require('webpack-stream');

const Instrumenter = isparta.Instrumenter;
const mochaGlobals = require('./test/setup/.globals');
const manifest = require('./package.json');

// Load all of our Gulp plugins
const $ = loadPlugins();

// Gather the library data from `package.json`
const config = manifest.babelBoilerplateOptions;
const mainFile = manifest.main;
const destinationFolder = path.dirname(mainFile);
const exportFileName = path.basename(mainFile, path.extname(mainFile));

function cleanDist(done) {
  del([destinationFolder]).then(() => done());
}

function cleanTmp(done) {
  del(['tmp']).then(() => done());
}

// Lint a set of files
function lint(files) {
  return gulp.src(files)
    .pipe($.eslint())
    .pipe($.eslint.format())
    .pipe($.eslint.failAfterError());
}

function lintSrc() {
  return lint('src/**/*.js');
}

function lintTest() {
  return lint('test/**/*.js');
}

function lintGulpfile() {
  return lint('gulpfile.js');
}

function build() {
  return gulp.src(path.join('src', config.entryFileName))
    .pipe(webpackStream({
      output: {
        filename: exportFileName + '.js',
        libraryTarget: 'umd',
        library: config.mainVarName
      },
      // Add your own externals here. For instance,
      // {
      //   jquery: true
      // }
      // would externalize the `jquery` module.
      externals: {},
      module: {
        loaders: [
          { test: /\.js$/, exclude: /node_modules/, loader: 'babel-loader' }
        ]
      },
      devtool: 'source-map'
    }))
    .pipe(gulp.dest(destinationFolder))
    .pipe($.filter(['**', '!**/*.js.map']))
    .pipe($.rename(exportFileName + '.min.js'))
    .pipe($.sourcemaps.init({ loadMaps: true }))
    .pipe($.uglify())
    .pipe($.sourcemaps.write('./'))
    .pipe(gulp.dest(destinationFolder));
}

function _mocha() {
  return gulp.src(['test/setup/node.js', 'test/unit/**/*.js'], {read: false})
    .pipe($.mocha({
      reporter: 'dot',
      globals: Object.keys(mochaGlobals.globals),
      ignoreLeaks: false
    }));
}

function _registerBabel() {
  require('babel-register');
}

function test() {
  _registerBabel();
  return _mocha();
}

function coverage(done) {
  _registerBabel();
  gulp.src(['src/**/*.js'])
    .pipe($.istanbul({
      instrumenter: Instrumenter,
      includeUntested: true
    }))
    .pipe($.istanbul.hookRequire())
    .on('finish', () => {
      return test()
        .pipe($.istanbul.writeReports())
        .on('end', done);
    });
}

const watchFiles = ['src/**/*', 'test/**/*', 'package.json', '**/.eslintrc'];

// Run the headless unit tests as you make changes.
function watch() {
  gulp.watch(watchFiles, ['test']);
}

function testBrowser() {
  // Our testing bundle is made up of our unit tests, which
  // should individually load up pieces of our application.
  // We also include the browser setup file.
  const testFiles = glob.sync('./test/unit/**/*.js');
  const allFiles = ['./test/setup/browser.js'].concat(testFiles);

  // Lets us differentiate between the first build and subsequent builds
  var firstBuild = true;

  // This empty stream might seem like a hack, but we need to specify all of our files through
  // the `entry` option of webpack. Otherwise, it ignores whatever file(s) are placed in here.
  return gulp.src('')
    .pipe($.plumber())
    .pipe(webpackStream({
      watch: true,
      entry: allFiles,
      output: {
        filename: '__spec-build.js'
      },
      // Externals isn't necessary here since these are for tests.
      module: {
        loaders: [
          // This is what allows us to author in future JavaScript
          { test: /\.js$/, exclude: /node_modules/, loader: 'babel-loader' },
          // This allows the test setup scripts to load `package.json`
          { test: /\.json$/, exclude: /node_modules/, loader: 'json-loader' }
        ]
      },
      plugins: [
        // By default, webpack does `n=>n` compilation with entry files. This concatenates
        // them into a single chunk.
        new webpack.optimize.LimitChunkCountPlugin({ maxChunks: 1 })
      ],
      devtool: 'inline-source-map'
    }, null, function() {
      if (firstBuild) {
        $.livereload.listen({port: 35729, host: 'localhost', start: true});
        var watcher = gulp.watch(watchFiles, ['lint']);
      } else {
        $.livereload.reload('./tmp/__spec-build.js');
      }
      firstBuild = false;
    }))
    .pipe(gulp.dest('./tmp'));
}

// Remove the built files
gulp.task('clean', cleanDist);

// Remove our temporary files
gulp.task('clean-tmp', cleanTmp);

// Lint our source code
gulp.task('lint-src', lintSrc);

// Lint our test code
gulp.task('lint-test', lintTest);

// Lint this file
gulp.task('lint-gulpfile', lintGulpfile);

// Lint everything
gulp.task('lint', ['lint-src', 'lint-test', 'lint-gulpfile']);

// Build two versions of the library
gulp.task('build', ['lint', 'clean'], build);

// Lint and run our tests
gulp.task('test', ['lint'], test);

// Set up coverage and run tests
gulp.task('coverage', ['lint'], coverage);

// Set up a livereload environment for our spec runner `test/runner.html`
gulp.task('test-browser', ['lint', 'clean-tmp'], testBrowser);

// Run the headless unit tests as you make changes.
gulp.task('watch', watch);

// An alias of test
gulp.task('default', ['test']);


src/serverless-alexa-plugin.js
import { merge } from 'lodash';

class ServerlessAlexaPlugin {
  constructor(serverless) {
    this.serverless = serverless;
    this.provider = 'aws';

    this.hooks = {
      'deploy:compileEvents': this.compileAlexaEvents.bind(this),
    };
  }

  compileAlexaEvents() {
    if (!this.serverless.service.resources.Resources) {
      throw new this.serverless.classes
        .Error('This plugin needs access to Resources section of the AWS CloudFormation template');
    }

    this.serverless.service.getAllFunctions().forEach((functionName) => {
      const functionObj = this.serverless.service.getFunction(functionName);

      if (functionObj.events) {
        for (let i = 0; i < functionObj.events.length; i++) {
          const event = functionObj.events[i];
          if (event === 'alexaSkillsKit' || event.alexaSmartHome) {
            const permissionTemplate = {
              Type: 'AWS::Lambda::Permission',
              Properties: {
                FunctionName: { 'Fn::GetAtt': [functionName, 'Arn'] },
                Action: 'lambda:InvokeFunction',
              },
            };

            if (event === 'alexaSkillsKit') {
              permissionTemplate.Properties.Principal = 'alexa-appkit.amazon.com';
            } else {
              if (typeof event.alexaSmartHome !== 'string') {
                const errorMessage = [
                  `Alexa Smart Home event of function ${functionName} is not a string`,
                  ' The correct syntax requires your skill\'s application ID from the',
                  ' Alexa Developer Console, example:',
                  ' alexaSmartHome: amzn1.ask.skill.12345678-1234-4234-8234-9234567890AB',
                  ' Please check the docs for more info.',
                ].join('');
                throw new this.serverless.classes
                  .Error(errorMessage);
              }
              permissionTemplate.Properties.Principal = 'alexa-connectedhome.amazon.com';
              permissionTemplate.Properties.EventSourceToken = event.alexaSmartHome;
            }

            const newPermissionObject = {
              [`${functionName}AlexaEventPermission${i}`]: permissionTemplate,
            };

            merge(this.serverless.service.resources.Resources,
              newPermissionObject);
          }
        }
      }
    });
  }
}

// TODO: consider `export default ServerlessAlexaPlugin;`
module.exports = ServerlessAlexaPlugin;
